File: 0200. Number of Islands/200_bfs.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return 0

        moves = [(-1, 0), (0, 1), (1, 0), (0, -1)]
        ans, row, col = 0, len(grid), len(grid[0])
        for i in range(row):
            for j in range(col):
                if grid[i][j] == '0':
                    continue

                if grid[i][j] == '1':
                    ans += 1
                    queue = [(i, j)]
                    while queue:
                        ti, tj = queue[-1]
                        queue.pop()
                        if ti < 0 or ti >= row or tj < 0 or tj >= col or grid[ti][tj] != '1':
                            continue

                        grid[ti][tj] = '0'
                        for move in moves:
                            queue = [(ti + move[0], tj + move[1])] + queue

        return ans

File: 0200. Number of Islands/200_bfs.cpp
class Solution {
public:
    struct pts{
        int r;
        int c;
    };
    int numIslands(vector<vector<char>>& grid) {
        queue<pts> que;
        int nIslands = 0;
        int m = grid.size(), n = grid[0].size();
        const vector<vector<int>> dirs = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
        for(int i=0; i<m; ++i){
            for(int j=0; j<n; ++j){
                if(grid[i][j]=='1'){
                    nIslands++;
                    que.push({i, j});
                    grid[i][j] = '0';
                    while(!que.empty()){
                        pts p = que.front();
                        que.pop();
                        for(auto& dir: dirs){
                            int nr = p.r + dir[0], nc = p.c + dir[1];
                            if(nr<0 || nr>=m || nc<0 || nc>=n) continue;
                            if(grid[nr][nc]=='1'){
                                que.push({nr, nc});
                                grid[nr][nc] = '0';
                            }
                        }
                    }
                }
            }
        }
        return nIslands;
    }
};
